compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -1067,6 +1070,192 @@ class LoweringConfigGenerator {
   SmallVector<bool> vectorScalableFlags;
 };
 
+/// A helper class that tracks dimension mappings both within individual
+/// operations and across multiple operations by analyzing the producer-consumer
+/// relationships of SSA values. This tracking is established by assigning a
+/// global dimension index to all loop dimensions encountered. Dimensions
+/// sharing the same global index are considered equivalent.
+class IterationDimTracker {
+public:
+  explicit IterationDimTracker(ArrayRef<Operation *> operations)
+      : operations(operations.begin(), operations.end()) {
+    // Ensure operations are processed in topological order.
+    mlir::computeTopologicalSorting(this->operations);
+    buildDimMapping();
+  }
+
+  /// Checks if the given dimension is common across all operations.
+  bool isCommonDim(Operation *op, unsigned pos) {
+    unsigned dim = operationToGlobalDimMaps[op][pos];
+    for (auto &[_, dims] : operationToGlobalDimMaps) {
+      if (!llvm::is_contained(dims, dim)) {
+        return false;
+      }
+    }
+    return true;
+  }
```

Comment:
I'd add an assertion, because the error message is more readable. Otherwise, it initializes an empty list and triggers an assertion when it accesses the `pos`-th element in the empty list.

```
assert(operationToGlobalDimMaps.contains(op));
```

btw, I think we can add `[[maybe_unused]]` in front of `auto` because `_` is not used and there may be warnings in debug build.